Lib/ast.py

```diff
@@ -1053,15 +1054,82 @@ def visit_AsyncWith(self, node):
         with self.block(extra=self.get_type_comment(node)):
             self.traverse(node.body)
 
+    def _str_literal_helper(
+        self, string, quote_types=("'", '"', '"""', "'''"), escape=""
+    ):
+        """Helper for writing string literals, minimizing escapes.
+        Returns (possible quote types, string literal to write).
+        """
+        # Escape characters we've been told to escape, backslashes, and
+        # non-printable characters other than \n and \t.
+        escape = {*escape, '\\'}
+        escaped_string = "".join(
+            (
+                c.encode('unicode_escape').decode('ascii')
+                if c in escape or (not c.isprintable() and c not in '\n\t')
+                else c
+            )
+            for c in string
+        )
+        possible_quotes = [q for q in quote_types if q not in escaped_string]
+        if "\n" in escaped_string:
+            possible_quotes = [q for q in possible_quotes if q in ('"""', "'''")]
+        if not possible_quotes:
+            # If there aren't any possible_quotes, fallback to using repr
+            # on the original string. Try to use a quote from quote_types.
+            string = repr(string)
+            quote = next((q for q in quote_types if string[0] in q), string[0])
+            return string[1:-1], [quote]
```

Comment:
This `next()` is a bit confusing. Is it for using the triple-quote for of whether double or single quote used? If so, I guess we can refactor it to something a bit more simpler?
```diff
+            original = repr(string)
+            quote_ty = original[0]
+            return original[1:-1], [quote_ty * 3 if is_multi_line else quote_ty]
```

Reply:
This doesn't quite do what we want. E.g., we use this code for docstrings, which we want to be triple quotes even if there aren't newlines. I added that to the comment though :-)